File: Sprint4/B/C++/main.cpp
// https://contest.yandex.ru/contest/23991/problems/B/
#include <iostream>
#include <unordered_map>
#include <random>
#include <string>

long long poly_hash(int a, int m, std::string& s) {
    long long h = 0, power = 1;
    int n = int(s.length());
    for (int i = n - 1; i >= 0; --i) {
        h = (h + int(s[i]) * power) % m;
        power = (power * a) % m;
    }
    return h;
}

std::string generate_random_string(int length) {
    static const char alphanum[] =
            "abcdefghijklmnopqrstuvwxyz";
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(0, sizeof(alphanum) - 2);

    std::string result;
    for (int i = 0; i < length; ++i) {
        result.push_back(alphanum[dis(gen)]);
    }
    return result;
}

int main() {
    int a = 1000;
    int m = 123987123;
    std::unordered_map<long long, std::string> hashes;

    int string_length = 20;
    while (true) {
        std::string s = generate_random_string(string_length);
        long long h = poly_hash(a, m, s);

        if (hashes.find(h) != hashes.end()) {
            std::cout << "Found two strings with the same hash!" << std::endl;
            std::cout << "String 1: " << hashes[h] << std::endl;
            std::cout << "String 2: " << s << std::endl;
            std::cout << "Hash: " << h << std::endl;
            break;
        }

        hashes[h] = s;
    }
    return 0;
}

File: Sprint4/B/Rust/src/main.rs
// https://contest.yandex.ru/contest/23991/problems/B/
use rand::{distributions::Distribution, distributions::Uniform};
use std::collections::HashMap;

fn poly_hash(a: i64, m: i64, s: &str) -> i64 {
    let mut h: i64 = 0;
    let mut power: i64 = 1;
    for c in s.chars().rev() {
        h = (h + (c as i64) * power) % m;
        power = (power * a) % m;
    }
    h
}

fn generate_random_string(length: usize) -> String {
    let alphanum: &[u8] = b"abcdefghijklmnopqrstuvwxyz";
    let mut rng = rand::thread_rng();
    let range = Uniform::from(0..alphanum.len());

    (0..length)
        .map(|_| alphanum[range.sample(&mut rng)] as char)
        .collect()
}

fn main() {
    let a: i64 = 1000;
    let m: i64 = 123987123;
    let mut hashes: HashMap<i64, String> = HashMap::new();

    let string_length: usize = 20;
    loop {
        let s = generate_random_string(string_length);
        let h = poly_hash(a, m, &s);

        if let Some(existing_string) = hashes.get(&h) {
            println!("Found two strings with the same hash!");
            println!("String 1: {}", existing_string);
            println!("String 2: {}", s);
            println!("Hash: {}", h);
            break;
        }

        hashes.insert(h, s);
    }
}
